### PR TITLE
Issue #1 - Allow custom themes

### PR DESCRIPTION
**Pull request by brittonhayes**
### Changes Introduced
- Adds support for defining theme in the header of a slideshow
- Support for default,dark, and light themes
- Adds theme header type and helper functions to read theme name but allow for more metadata if you'd like to add it later

The themes can be specified like this:

```markdown
---
theme: "default"
---
```
```markdown
---
theme: "dark"
---
```
```markdown
---
theme: "light"
---
```

Adds support for issue https://github.com/maaslalani/slides/issues/1 with custom user themes

### CODE REVIEW

**Comment by maaslalani** [comment]
Hey! Thanks so much for contributing to slides. This looks good and I would be happy to add this to the project. However, one other way I was thinking (instead of a command line option) would be to have the config live directly inside the presentations.

For example:
```markdown
---
theme: styles.json
---

# Slide 1
Stuff here

---

# Slide 2
More stuff here
```

This way we can keep the command line flags super minimal and lean towards placing the configuration inside the `slides.md`.
So that the user only needs to write
```
slides presentation.md
```

rather than:
```
slides --theme styles.json presentation.md
```
everytime.

What do you think?

**Comment by brittonhayes** [comment]
I think that's a fun idea! 

Another expansion onto that I feel like would be the ultimate is allow a minimal set of parameters for users to theme their slides directly in the markdown file. 

You could leverage something sort of like Hugo frontmatter to specify your basic stuff at the head of the markdown file. That way people can exchange slides with a single markdown file and no need to share a separate `theme.json`.

Some references to that:

- https://gohugo.io/content-management/front-matter/#front-matter-formats
- https://github.com/gohugoio/hugo/blob/master/parser/frontmatter.go

Totally down to help pivot it whichever way you see fit though :smile:  This is your project! I'm just here to be a friendly passerby and help ya accomplish what you're lookin to do!

**Comment by maaslalani** [comment]
Ohhhh! Yeah, that could be great!

I think to start we can simply have the theme key as a config option:
We can simply forward this option to the [glamour](https://github.com/charmbracelet/glamour) renderer that we use.

I believe `glamour` accepts any JSON file (path), as well as options like: `ascii`, `dark`, and `light` so we will also get those themes for free if we just pass the value along. By default we should use our `theme.json`.

**Comment by maaslalani** [comment]
I think it would be also dope to have a bunch of preset slide themes that users can submit and we can have them in this repo so that they come preloaded and users don't have to share the `theme.json` either.
```markdown
---
theme: custom-user-theme.json
---
```

That might bloat the binary size, if we have too many though. So potentially a better solution would be to allow full URL links to json files that we could load on demand? Could be a fun idea:

```markdown
---
theme: https://raw.githubusercontent.com/maaslalani/slides/main/styles/theme.json
---
```

**Comment by brittonhayes** [comment]
> `theme: [filepath.json|ascii|light|dark]`

Honestly, at the end of the day I think this is a great starting point and would serve most people's initial needs. Simpler to implement and lets people choose their flavor without having to think too hard about it.

Separate question: How are you breaking up each `---` into a new page? I'm trying to find it in the repo but cannot for the life of me figure out where that is being performed. Since I'm assuming this is how we'd add the ability to parse the model from the "first page".

**Comment by maaslalani** [comment]
The action is here: https://github.com/maaslalani/slides/blob/main/cmd/root.go#L43

The delimiters are defined above. I assume we would check for if the document starts with a `---` and consume until the next `---` and assume that is all valid configuration, I'd suggest we don't fail too loudly if the configuration is invalid (since the user will probably be able to tell)

**Comment by brittonhayes** [comment]
Alrighty, I've added support for the ability to specify the following themes in the front matter. I haven't made any changes to these theme.json files, since I figured you'd wanna be in control of the aesthetic :smile:.

Also, I didn't add the ability to specify a custom theme.json file since I wanted to get a feel for how you feel about this change before implementing that.

```markdown
---
theme: "default"
---
```
```markdown
---
theme: "dark"
---
```
```markdown
---
theme: "light"
---
```

**Comment by maaslalani** [comment]
Nice this is looking great! Also we don't need to define `theme_dark.json` or `theme_dark.json` because these are already defined in https://github.com/charmbracelet/glamour/tree/master/styles/gallery and we can just pass the `"dark"`, `"light"` string to glamour.

File: cmd/root.go [hunk 1]
```diff
@@ -41,17 +43,26 @@ var root = &cobra.Command{
 		content := string(b)
 		content = strings.ReplaceAll(content, altDelimiter, delimiter)
 		slides := strings.Split(content, delimiter)
+		m, err := meta.New().ParseHeader(slides[0])
+		if err != nil {
+			return errors.New("could not parse theme from header")
```
Comment (maaslalani) [inline cmd/root.go:48]: I think instead of `error`ing out we should simply use the default theme

Comment (brittonhayes) [reply cmd/root.go:48]: Was just thinking that after reading back through. Switching it now.

**Comment by brittonhayes** [comment]
Updated the theme selector so now all of the following theme names are valid for Slides.

```go
// SelectTheme picks a glamour style config based
// on the theme provided in the markdown header
func SelectTheme(theme string) ansi.StyleConfig {
	switch theme {
	case "ascii":
		return glamour.ASCIIStyleConfig
	case "light":
		return glamour.LightStyleConfig
	case "notty":
		return glamour.NoTTYStyleConfig
	default:
		return glamour.DarkStyleConfig
	}
}
```

**Comment by maaslalani** [comment]
This is looking super great! Coming together nicely!

So we have a couple options:
- We can initialize the glamour renderer in the root.go and then store a reference to it in the model
- We can also probably contribute this theme to glamour and then see what they say 😄 (this would probably be the cleanest option)
- Since glamour can take a string for the path of a theme we can create a file in $HOME/.config/slides/theme.json and then the default behaviour would be for glamour to use this path and then a use could also simply write a different path for the theme

**Comment by brittonhayes** [comment]
I kinda dig the idea of seeing if it could be added to the glamour set of themes but I'm not sure how long that would take to get a green light from them 🤔 

The happiest path would be if the theme was defined as an ascii.StyleConfig or if you could load an ascii.StyleConfig from a filepath rather than only a termrenderer option. That way we could keep that super clean switch statement but also support custom themes from the file system.

Maybe that's a change worth adding to glamour??

**Comment by maaslalani** [comment]
Yea it would be nice to have the theme in glamour but I think it might not be good because if we ever want to update it, we will first need to update glamour, cut a glamour release, then update the dependencies in slides then cut a slides release. So maybe not the best option for now.

Personally, I think the best bet is to initialize the glamour renderer in root and store it in the model

**Comment by brittonhayes** [comment]
I'm cool with that! Lemme take a crack at it real quick and let me know if looks like we're on the same page.

**Comment by brittonhayes** [comment]
Alllrighty. You can now specify a theme with any of the following options, each of them should fall back to a default to prevent any nasty error messages.

```markdown
---
theme: "default"
---
```
```markdown
---
theme: "dark"
---
```
```markdown
---
theme: "light"
---
```
```markdown
---
theme: "somefolder/theme.json"
---
```

**Comment by brittonhayes** [comment]
@maaslalani Sorry, just realized you made some changes on the main branch and I got confused :joy: I should really sleep more. I'll fix mine up so it doesn't conflict.

**Comment by NavTheNugget** [comment]
Themes! Let's go!

**Comment by maaslalani** [comment]
@brittonhayes Just confirming that if the user doesn't have a frontmatter config that the slides will still work right? I want to make sure everything is completely optional.

**Comment by brittonhayes** [comment]
> @brittonhayes Just confirming that if the user doesn't have a frontmatter config that the slides will still work right? I want to make sure everything is completely optional.

Good callout! Now it supports that! :smile: I added a couple tests to ensure this is the case.

- If an user specifies no theme, it will default to your embedded DefaultTheme and treat the first slide as part of the slideshow. 
- If an user provides a custom theme, it will treat the first slide as frontmatter, apply their specified theme, and show `slides[1:]` for the slideshow.
- If an user provides a custom theme but it isn't a valid theme name or a valid filepath, it will use the glamour.DarkStyleConfig, since that is set to the default for custom themes.

I'm just writing a couple more tests to cover some of the more unique use-cases to make sure I'm not incorrect here!

**Comment by maaslalani** [comment]
Those cases look correct to me. And thank you so so so much for writing tests! Makes me so happy to see tests haha, totally not necessary but very much appreciated ❤️

I'll probably end up writing a full test suite and some point, which should be fun! 😄

**Comment by maaslalani** [comment]
Totally optional, but if you want to ship this github action https://github.com/maaslalani/slides/commit/cf7053ce1bca39650c65f290647180ed22cff6d1 as a separate PR with a dummy test would be happy to merge that quickly

**Comment by brittonhayes** [comment]
> Totally optional, but if you want to ship this github action [cf7053c](https://github.com/maaslalani/slides/commit/cf7053ce1bca39650c65f290647180ed22cff6d1) as a separate PR with a dummy test would be happy to merge that quickly

Yeah I can do that real quick!

**Comment by brittonhayes** [comment]
@maaslalani Want me to push a dummy change here just to see if all the tests and lints pass for the hell of it?

**Comment by maaslalani** [comment]
If you rebase and force push it should do that

**Comment by brittonhayes** [comment]
Uhhh woopssss. I definitely just messed up my rebase and didn't mean to close the PR :joy:

**Comment by maaslalani** [comment]
Oh no! Looks like a bad rebase or something. [`git reflog`](https://stackoverflow.com/questions/134882/undoing-a-git-rebase) can hopefully help.

**Comment by maaslalani** [comment]
> Uhhh woopssss. I definitely just messed up my rebase and didn't mean to close the PR :joy:

Yea haha I figured! 😂😂

**Pull request closed without merging**